File: Data Structures & Algorithms/validate-parentheses/submission-13.py
class Solution:
    def isValid(self, s: str) -> bool:
        pair = {"}":"{", ")":"(", "]":"["}
        stack = []
        for ch in s:
            if ch not in pair:
                stack.append(ch)
            else:
                if stack.__len__() == 0 or stack[-1] != pair[ch]:
                    return False
                else:
                    stack.pop()
        return stack.__len__() == 0

File: Data Structures & Algorithms/course-schedule-ii/submission-13.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        vector<vector<int>> graph(numCourses, vector<int>());
        vector<int> result;
        for (const auto& p : prerequisites) {
            indegree[p[1]]++;
            graph[p[0]].push_back(p[1]);
        }
        queue<int> que;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                que.push(i);
            }
        }
        while (!que.empty()) {
            int curr = que.front();
            que.pop();
            result.push_back(curr);
            for (const int preq : graph[curr]) {
                indegree[preq]--;
                if (indegree[preq] == 0) {
                    que.push(preq);
                }
            }
        }
        if ((int)result.size() != numCourses) {
            return {};
        }
        reverse(result.begin(), result.end());
        return result;
    }
};
